contracts/lsm-share-bond-provider/src/tests.rs

```diff
@@ -0,0 +1,1584 @@
+use std::borrow::BorrowMut;
+
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{
+    attr, coins, from_json,
+    testing::{mock_env, mock_info, MockApi},
+    to_json_binary, Addr, Coin, Decimal, Decimal256, Event, MemoryStorage, OwnedDeps, Response,
+    SubMsg, Timestamp, Uint128,
+};
+use cw_ownable::{Action, Ownership};
+use cw_utils::PaymentError;
+use drop_helpers::{
+    ica::IcaState,
+    testing::{mock_dependencies, WasmMockQuerier},
+};
+use drop_staking_base::{
+    error::lsm_share_bond_provider::ContractError,
+    msg::puppeteer::DelegationsResponse,
+    state::{
+        lsm_share_bond_provider::{
+            Config, ConfigOptional, ReplyMsg, TxState, CONFIG, LAST_LSM_REDEEM, PENDING_LSM_SHARES,
+            TOTAL_LSM_SHARES_REAL_AMOUNT, TX_STATE,
+        },
+        puppeteer::{Delegations, DropDelegation},
+    },
+};
+use neutron_sdk::{
+    bindings::{
+        msg::{IbcFee, NeutronMsg},
+        query::NeutronQuery,
+    },
+    query::min_ibc_fee::MinIbcFeeResponse,
+    sudo::msg::RequestPacketTimeoutHeight,
+};
+
+use crate::contract::check_denom::{DenomTrace, QueryDenomTraceResponse};
+
+use prost::Message;
+
+fn get_default_config(lsm_redeem_threshold: u64, lsm_redeem_maximum_interval: u64) -> Config {
+    Config {
+        puppeteer_contract: Addr::unchecked("puppeteer_contract"),
+        core_contract: Addr::unchecked("core_contract"),
+        validators_set_contract: Addr::unchecked("validators_set_contract"),
+        port_id: "port_id".to_string(),
+        transfer_channel_id: "transfer_channel_id".to_string(),
+        timeout: 100u64,
+        lsm_min_bond_amount: 100u128.into(),
+        lsm_redeem_threshold,
+        lsm_redeem_maximum_interval,
+    }
+}
+
+#[cw_serde]
+pub struct QueryDenomTraceRequest {
+    pub hash: String,
+}
+
+fn lsm_denom_query_config(
+    deps: &mut OwnedDeps<MemoryStorage, MockApi, WasmMockQuerier, NeutronQuery>,
+    unknown_validator: bool,
+) {
+    deps.querier.add_stargate_query_response(
+        "/ibc.applications.transfer.v1.Query/DenomTrace",
+        |request| {
+            let request =
+                cosmos_sdk_proto::ibc::applications::transfer::v1::QueryDenomTraceRequest::decode(
+                    request.as_slice(),
+                )
+                .unwrap();
+            if request.hash == "lsm_denom_1" {
+                to_json_binary(&QueryDenomTraceResponse {
+                    denom_trace: DenomTrace {
+                        base_denom: "valoper12345/1".to_string(),
+                        path: "transfer/transfer_channel_id".to_string(),
+                    },
+                })
+                .unwrap()
+            } else {
+                to_json_binary(&QueryDenomTraceResponse {
+                    denom_trace: DenomTrace {
+                        base_denom: "valoper12345/1".to_string(),
+                        path: "wrong_denom".to_string(),
```

Comment:
This is not how this module works. Path will always contain valid data. If you want to check for errors, please mock an error instead, not malformed data.

Reply:
No, the idea here is to simulate exception `InvalidDenom` in the `check_denom`, this was the simpliest way

Reply:
Okay, since check_denom is prepared to handle invalid data, then this mock makes sense